File: agent/web/mutation_guard.py
from __future__ import annotations

from typing import Any

from aiohttp import web

from ..control import ControlMode


def assert_web_mutation_allowed(
    control: dict[str, Any],
    *,
    area: str,
    action: str,
) -> None:
    """Apply WebUI mutation semantics that sit outside agent tool approvals."""
    pending = control.get("pending") if isinstance(control.get("pending"), dict) else None
    if pending:
        raise web.HTTPConflict(
            reason=(
                f"Cannot {action} {area} while Ask / Plan is pending; "
                "answer, approve or cancel the pending interaction first."
            )
        )
    if str(control.get("mode") or "") == ControlMode.PLAN.value:
        raise web.HTTPForbidden(
            reason=(
                f"Cannot {action} {area} in plan mode; approve or leave Plan mode "
                "before executing mutations."
            )
        )

File: packages/core/src/api/mutation-guard.ts
export class CoreMutationGuardError extends Error {
  readonly status: 403 | 409

  constructor(status: 403 | 409, message: string) {
    super(message)
    this.name = 'CoreMutationGuardError'
    this.status = status
  }
}

export function assertCoreMutationAllowed(
  control: Record<string, unknown>,
  opts: { area: string; action: string },
): void {
  const pending = isRecord(control.pending) ? control.pending : null
  if (pending) {
    throw new CoreMutationGuardError(
      409,
      `Cannot ${opts.action} ${opts.area} while Ask / Plan is pending; answer, approve or cancel the pending interaction first.`,
    )
  }
  if (String(control.mode || '') === 'plan') {
    throw new CoreMutationGuardError(
      403,
      `Cannot ${opts.action} ${opts.area} in plan mode; approve or leave Plan mode before executing mutations.`,
    )
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value))
}
